components/Blueprints/Runner.php

```diff
@@ -402,127 +403,22 @@ private function validateBlueprint(): void {
-				}
-			} else {
-				if ( isset( $wp_version['min'] ) ) {
-					if ( $wp_version['min'] === 'latest' ) {
-						throw new BlueprintExecutionException(
-							'Setting wordpressVersion.min to "latest" is not allowed and probably not what you want. Either set wordPressVersion.recommended to "latest" or set wordPressVersion.min to a specific version string instead.'
-						);
-					}
-					$min = WordPressVersion::fromString( $wp_version['min'] );
-					if ( ! $min ) {
-						throw new BlueprintExecutionException( 'Invalid WordPress version string in wordpressVersion.min: ' . $wp_version['min'] );
-					}
-				}
-				// Latest version is implicitly the default and it's only for resolving
-				// the WordPress version to install. It's not used for version checks on
-				// existing sites and VersionConstraint doesn't support it. It doesn't have
-				// enough information anyway – the meaning of "latest" changes over time.
-				if ( isset( $wp_version['max'] ) && $wp_version['max'] !== 'latest' ) {
-					$this->recommendedWpVersion = $wp_version['max'];
-					$max = WordPressVersion::fromString( $wp_version['max'] );
-					if ( ! $max ) {
-						// @TODO: Reuse this error message
-						// 'Unrecognized WordPress version. Please use "latest", a URL, or a numeric version such as "6.2", "6.0.1", "6.2-beta1", or "6.2-RC1"'
-						throw new BlueprintExecutionException( 'Invalid WordPress version string in wordpressVersion.max: ' . $wp_version['max'] );
-					}
-				}
-				if ( isset( $wp_version['recommended'] ) && $wp_version['recommended'] !== 'latest' ) {
-					$this->recommendedWpVersion = $wp_version['recommended'];
-					$recommended = WordPressVersion::fromString( $wp_version['recommended'] );
-					if ( false === $recommended ) {
-						throw new BlueprintExecutionException( 'Invalid WordPress version string in wordpressVersion.recommended: ' . $wp_version['recommended'] );
-					}
-				}
-			}
-
-			$this->wpVersionConstraint = new VersionConstraint( $min, $max, $recommended );
-			$wpConstraintErrors        = $this->wpVersionConstraint->validate();
-			if ( ! empty( $wpConstraintErrors ) ) {
-				throw new BlueprintExecutionException( 'Invalid WordPress version constraint: ' . implode( '; ', $wpConstraintErrors ) );
-			}
-			// Note: In here's we're only checking if the version constraint is defined
-			// correctly. The actual version check for WordPress is done in
-			// NewSiteResolver and ExistingSiteResolver.
-		}
-
-		// Validate the override constraint if it was set
-		if ( $this->wpVersionConstraint ) {
-			$wpConstraintErrors = $this->wpVersionConstraint->validate();
-			if ( ! empty( $wpConstraintErrors ) ) {
-				throw new BlueprintExecutionException( 'Invalid WordPress version constraint from CLI override: ' . implode( '; ', $wpConstraintErrors ) );
-			}
+		$validator = new \WordPress\Blueprints\Validator\BlueprintValidator( $this->blueprintArray, $this->blueprintExecutionContext );
```

Comment:
Nice idea separating it! I've been experimenting with co-locating that logic with parsing the Blueprint and it had some ups and downs.

The downside here is that it's separated from parsing the Blueprint. The upside here is that it's separated from parsing the Blueprint. Let's roll with a separate validator and see how it goes.

Reply:
> The downside here is that it's separated from parsing the Blueprint. The upside here is that it's separated from parsing the Blueprint.

This. I think that if we can make the parsing logic collect all errors before bailing out while keeping a reasonably nice code, then we should go with that. I'm looking into that.